File: frontend/types/content.ts
import {Credits} from "./credits";
import {Providers} from "./providers";
import {Season} from "./season";

export interface ContentItem {
    id: number;
    title: string;
    genres: string[];
    type: ContentType;
    overview: string;
    length: number;
    rating: number;
    releaseDate: string;
    posterPath: string;
    backdropPath: string;
    trailerId?: string;
}

export type ContentType = "movie" | "tv";

export interface ContentItemDetails extends ContentItem {
    providers: Providers;
    credits: Credits;
    seasons?: Season[];
    similar?: ContentItem[];
}

File: backend/src/main/java/me/nextwatch/NextWatch/content/Content.java
package me.nextwatch.NextWatch.content;

import jakarta.annotation.Nullable;
import jakarta.persistence.*;
import lombok.*;
import me.nextwatch.NextWatch.content.credit.Credit;
import me.nextwatch.NextWatch.content.provider.Provider;
import me.nextwatch.NextWatch.content.season.Season;
import org.hibernate.annotations.Array;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.UpdateTimestamp;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.time.LocalDate;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Entity
public class Content {
    @EmbeddedId
    private ContentId id;

    private String title;
    private List<Integer> genreIds;
    private String overview;
    private Integer length;
    private Double rating;
    private String posterPath;
    private String backdropPath;
    private LocalDate releaseDate;
    private String trailerId;

    @CreationTimestamp
    private Instant createdAt;

    @UpdateTimestamp
    private Instant updatedAt;

    @Nullable
    @OneToMany(mappedBy = "content", cascade = CascadeType.ALL)
    private List<Season> seasons;

    @OneToMany(mappedBy = "content", cascade = CascadeType.ALL)
    private List<Credit> credits;

    @OneToMany(mappedBy = "content", cascade = CascadeType.ALL)
    private List<Provider> providers;

    @JdbcTypeCode(SqlTypes.VECTOR)
    @Array(length = 768)
    @Column(columnDefinition = "vector(768)")
    private float[] embedding;
}
